Enable x86 CPU vectorization on windows [submodule sleef]

xuhancn opened the pull request:
Enable VEC on Windows OS.
1. Fix some type defination gap between Windows and Linux.
2. Fix some operator not support on Windows, such as [], /.
3. Enable static sleef library build on Windows.
4. Disable unsupported function overloading on MSVC.
5. Upgrade submodule sleef lib, which fixed build issue on Windows.
6. Fixed bazel build issues.
7. Fix test app not link to sleef on Windows.

Note: If rebuild fail after pulled this PR, please sync `sleef` submodule by run:
```cmd
git submodule sync
git submodule update --init --recursive
```

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @frank-wei

jgong5 commented on aten/src/ATen/native/cpu/GridSamplerKernel.cpp:
```diff
@@ -578,13 +578,15 @@ struct ApplyGridSample<scalar_t, 2, GridSamplerInterpolation::Bilinear,
 
       auto interpolated = (nw_val * nw) + (ne_val * ne) + (sw_val * sw) + (se_val * se);
       interpolated.store(out_slice[c].data() + offset, len);
+      #endif
```
no else?

xuhancn replied:
It is WA during development, and fixed now.

jgong5 commented on aten/src/ATen/native/cpu/GridSamplerKernel.cpp:
```diff
@@ -683,13 +685,15 @@ struct ApplyGridSample<scalar_t, 2, GridSamplerInterpolation::Bilinear,
 
       gx = gx + ((ne_val - nw_val) * s + (se_val - sw_val) * n) * gOut;
       gy = gy + ((sw_val - nw_val) * e + (se_val - ne_val) * w) * gOut;
+      #endif
```
ditto

jgong5 commented on aten/src/ATen/native/cpu/GridSamplerKernel.cpp:
```diff
@@ -761,8 +765,10 @@ struct ApplyGridSample<scalar_t, 2, GridSamplerInterpolation::Nearest,
+      #ifndef _MSC_VER
       auto inp_val = mask_gather<sizeof(scalar_t)>(Vec(0), inp_slice_ptr, i_offset, mask_copy);
       inp_val.store(static_cast<void*>(out_ptr), len);
+      #endif
```
ditto

jgong5 commented on aten/src/ATen/native/cpu/GridSamplerKernel.cpp:
```diff
@@ -1131,13 +1140,15 @@ static inline void grid_sample_2d_grid_slice_iterator(
         grid_ptr_x += i_offset_delta;
         grid_ptr_y += i_offset_delta;
         spatial_offset += len;
+        #endif
```
ditto

ezyang commented:
@pytorchbot merge

xuhancn commented:
@pytorchbot label "topic: not user facing"

xuhancn commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -i

xuhancn commented:
@pytorchbot rebase

xuhancn commented:
@pytorchbot merge

xuhancn committed:
Enable x86 CPU vectorization on windows
* Fix some type defination gap between Windows and Linux.
* Fix some operator not support on Windows, such as [], /.
* Enable static sleef library build on Windows.
* Disable unsupported function overloading on MSVC.
* Update sleef to latest, fix .so share lib disappeared issue.
* Fix test app link issue.

xuhancn commented:
@pytorchbot rebase

ezyang commented:
Do you think the test_grad_scaling_unscale_cpu_float32 failure is real?

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -1526,6 +1526,35 @@ if(MSVC AND BUILD_SHARED_LIBS)
   target_compile_options(torch_cpu PRIVATE "-DONNX_BUILD_MAIN_LIB")
 endif()
 
+if(MSVC)
```
This should be guarded behind `USE_SLEEF`

malfet replied:
Also, do you mind moving this to (and just remove MSVC part) https://github.com/pytorch/pytorch/blob/5410385c427cf2969da4d08b8ea9e4ed4c6b82d5/aten/src/ATen/CMakeLists.txt#L423

xuhancn replied:
> This should be guarded behind `USE_SLEEF`

1. USE_SLEEF is CPP macro: https://github.com/pytorch/pytorch/blob/8c2fde1fcf17a1108557740d581e1808cc5b7d38/aten/src/ATen/cpu/vec/vec256/vec256_float_neon.h#L38-L42
2. CMAKE not contains USE_SLEEF switch. Sleef is hard dependency of torch_cpu, if vec enabled. The only option is whether use system sleef: https://github.com/pytorch/pytorch/blob/8c2fde1fcf17a1108557740d581e1808cc5b7d38/CMakeLists.txt#L424

xuhancn replied:
<img width="1353" alt="Image" src="https://github.com/pytorch/pytorch/assets/8433590/f8a1056b-339c-408c-ba17-aa1dd59bc65b">
I referenced to the "ONNX", which is linked to torch_cpu success. It was processed as submodule in caffee2/CMakeLists.txt.
So I think I can only process sleef here.

malfet reviewed (changes requested):
Overall LGTM, but please re-use as much cmake code as possible

xuhancn commented:
> Do you think the test_grad_scaling_unscale_cpu_float32 failure is real?

![image](https://github.com/pytorch/pytorch/assets/8433590/08e0fe05-b877-4a32-84f3-fb0b2b1bc9c3)
In my dev machine, this test is passed. 

And I trigger a dummy PR to check the origin case, but no Windows test was run. https://github.com/pytorch/pytorch/pull/119270

I guess it is not a real failure.

xuhancn commented:
> > Do you think the test_grad_scaling_unscale_cpu_float32 failure is real?
> 
> ![image](https://private-user-images.githubusercontent.com/8433590/302852232-08e0fe05-b877-4a32-84f3-fb0b2b1bc9c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyODg5NTMsIm5iZiI6MTcwNzI4ODY1MywicGF0aCI6Ii84NDMzNTkwLzMwMjg1MjIzMi0wOGUwZmUwNS1iODc3LTRhMzItODRmMy1mYjBiMmIxYmM5YzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDdUMDY1MDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzM2NDZlZWQ3MWM4ZjAzNTdiZjc2ZGNiZjM4M2IzMmU2OWIwNmMxMTc5ZjA0NGYyYjFiNTY2NDg1MjlkNDJiOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.R8x-BKyL9zkcGI1wGYNfuikW1zSF0okuLJiRasqWghU) In my dev machine, this test is passed.
> 
> And I trigger a dummy PR to check the origin case, but no Windows test was run. #119270
> 
> I guess it is not a real failure.

Installed hypothesis, it passed also.
![image](https://github.com/pytorch/pytorch/assets/8433590/a2c5b821-67b6-471b-a5b3-dc76df24fb2a)

xuhancn commented:
@pytorchbot rebase

malfet reviewed (approved):
Lets land it to avoid code getting stale and do some build followups later

malfet commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'This is breaking Window binary build https://github.com/pytorch/pytorch/actions/runs/7874475000/job/21484997298 where it failed to build sleef' -c nosignal

huydhn commented:
Only debug build is failing from what I see

huydhn commented:
@pytorchbot rebase -b main

xuhancn commented:
It depends on https://github.com/shibatch/sleef/pull/510.

xuhancn commented:
Sleef is suspend to merge new PR due to release: https://github.com/shibatch/sleef/discussions/498

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
Status update:
1. We still pending on sleef PR can't merge: https://github.com/shibatch/sleef/pull/510 , maintainer said that is limited bandwidth to review and merge code: https://github.com/shibatch/sleef/discussions/498#discussioncomment-8527530
2. I had to fork a sleef repo to move on my develop work. The forked repo contains my new sleef PR.
3. Current PR almost works on my forked sleef.
4. Still two bazel fail, and they are should be related to original sleef. It is not diffcute issue.
5. This PR can boost performance on all Windows machine, I wish we can catch up pytorch 2.3 release.
CC: @jgong5 @malfet @ezyang @blapie

ezyang commented on .gitmodules:
```diff
@@ -149,3 +149,6 @@
 	url = https://github.com/microsoft/mimalloc.git
+[submodule "third_party/sleef_fork"]
+	path = third_party/sleef_fork
+	url = https://github.com/xuhancn/sleef.git
```
@malfet idk if you have an opinion about this, but in the past when we've needed to temporarily fork the repo, we typically updated the url in the submodule itself. For example, https://github.com/pytorch/pytorch/pull/48120

xuhancn replied:
Done, for test.

ezyang commented on third_party/sleef_fork:
The commit history is a bit messy. For ease of understanding the change here, it would be best if this was (1) the exact version of sleef we were previously on + (2) one patch with all of the changes for Windows compatibility. We then can update sleef by rebasing the patch in question.

xuhancn replied:
rebase to one commit for code.

ezyang commented:
@xuhancn for speed of integration, may I recommend splitting the C++ changes (which hopefully we can just merge immediately) from the build system changes + sleef submodule changes?

xuhancn commented:
> @xuhancn for speed of integration, may I recommend splitting the C++ changes (which hopefully we can just merge immediately) from the build system changes + sleef submodule changes?

1. Need Meta approve the fork sleef repo.
2. C++ change will turn on sleef, and it can't build pass in Windows without sleef update.
3. Due to current sleef was two years ago version, and build options changed. Can't seprate sleef and C++.

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
@ezyang  Current status:
1. Rebased commits into two commit: cpp related and sleef related.
2. Switch sleef to temp repo via URL.
3. Fixed bazel build fail.
This PR can pass all preCI test now.

Due to this PR will switch sleef repo to my temp repo, and it contains the necessary PR(https://github.com/shibatch/sleef/pull/510) is not merged.
Whether merge this PR, it up to Meta engineers. @ezyang @malfet

xuhancn commented:
Add a locked pytorch dedicate branch: https://github.com/xuhancn/sleef/tree/pytorch_release

xuhancn commented:
SLEEF will merge this PR(https://github.com/shibatch/sleef/pull/531) to fix the build issue. Let's wait for a while.

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot merge -i

blapie commented:
Have you tried with the most recent version of SLEEF? 
The .so disappeared due to the previous fix but got put back in this commit https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b

xuhancn commented:
> Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)

I will try latest sleef.

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
> Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)

@zou3519 retry to latest sleef. It should fixed your problem.

xuhancn commented:
@pytorchmergebot merge -i

zou3519 commented:
> > Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)
> 
> @zou3519 retry to latest sleef. It should fixed your problem.

How does one "retry to latest sleef"? The thing I do when installing PyTorch is checking out all submodules. Is that sufficient or do I need to install something new?

xuhancn commented:
> > > Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)
> > 
> > 
> > @zou3519 retry to latest sleef. It should fixed your problem.
> 
> How does one "retry to latest sleef"? The thing I do when installing PyTorch is checking out all submodules. Is that sufficient or do I need to install something new?

Sleef seems some regression, The latest PR(https://github.com/shibatch/sleef/pull/535) looks like fix it.

xuhancn commented:
@zou3519 ping you at slack, free to discuss?

xuhancn commented:
@zou3519 may I known if you run the command update `sleef` submodule?
```cmd
git submodule sync
git submodule update --init --recursive
```

xuhancn commented:
Hi @zou3519 here two possible reason caused you build fail:
1. Sleef regression, the new sleef PR(https://github.com/shibatch/sleef/pull/535) fixed it. I have update the sleef to the fixed version.
2. You didn't update `sleef` submodule, please run below cmd:
```cmd
git submodule sync
git submodule update --init --recursive
```
This PR passed all of post CI, I some confidence to its quality.

xuhancn commented:
@pytorchmergebot merge -m "rebase all commits to one, update sleef. And then re-merge."

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot merge --rebase

xuhancn committed:
Merge branch 'pytorch:main' into xu_turn_on_vec_on_windows

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot merge -i

xuhancn commented:
> @pytorchbot revert -m "This caused build failures for 2+ pytorch devs, so we're reverting it to be safe" -c weird

Please sync `sleef` submodule by run:
```cmd
git submodule sync
git submodule update --init --recursive
```

xuhancn commented:
> Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)

@blapie merge this PR again, with latest sleef.

huydhn commented:
@pytorchbot revert -m 'Sorry for revert your change one more time but the hard part is that it breaks lot of internal builds' -c ghfirst

@ezyang @malfet As reviewers, could one of you help take a look at the failures on D55166757 to unblock this?  From what I see, the build error is `unknown type name 'SLEEF_CONST'`.  Do we have guide somewhere on how to handle external module update in fbcode?

xuhancn commented:
> @pytorchbot revert -m 'Sorry for revert your change one more time but the hard part is that it breaks lot of internal builds' -c ghfirst
> 
> @ezyang @malfet As reviewers, could one of you help take a look at the failures on D55166757 to unblock this? From what I see, the build error is `unknown type name 'SLEEF_CONST'`. Do we have guide somewhere on how to handle external module update in fbcode?

`SLEEF_CONST` Need upgrade to latest sleef lib. @huydhn original sleef can't working on pytorch Windows.

xuhancn committed:
fix unknown type name SLEEF_CONST issue.

xuhancn commented:
> > @pytorchbot revert -m 'Sorry for revert your change one more time but the hard part is that it breaks lot of internal builds' -c ghfirst
> > @ezyang @malfet As reviewers, could one of you help take a look at the failures on D55166757 to unblock this? From what I see, the build error is `unknown type name 'SLEEF_CONST'`. Do we have guide somewhere on how to handle external module update in fbcode?
> 
> `SLEEF_CONST` Need upgrade to latest sleef lib. @huydhn original sleef can't working on pytorch Windows.

@huydhn I have add some code to handle this issue: https://github.com/pytorch/pytorch/pull/118980/commits/aa83a5277b59651f75595de14bdaed8b3f4f2927 and re-merge it. Please help me on trace issue.

xuhancn commented:
@pytorchmergebot merge

huydhn commented:
Let me import your change first to see if it fixes the internal build failures, and I will merge this once I could confirm it

huydhn commented:
Another build failure shows up FYI:

```
aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:492:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_logf8_u10);
```

xuhancn committed:
compatiable to old version sleef.

xuhancn commented:
> Another build failure shows up FYI:
> 
> ```
> aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:492:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
>     return map(Sleef_logf8_u10);
> ```

Please import with it: https://github.com/pytorch/pytorch/pull/118980/commits/da32e86c1083892ee6c9a023f9f673d73b1b092b

xuhancn commented:
@pytorchmergebot rebase

xuhancn committed:
Merge branch 'pytorch:main' into xu_turn_on_vec_on_windows

xuhancn commented:
> @huydhn has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55208753).

@huydhn Any update?

huydhn commented:
Sorry for the delay, @malfet mentions that sleef module is already there internall on fbcode, so I think I could try to update it to block this change.

anijain2305 commented:
Not sure if related. I am on `main` branch (not on this PR) and seeing bulid failure like these

```
/data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h: In instantiation of ‘at::vec::AVX2::Vectorized<T> at::vec::AVX2::Vectorized16<T>::cosh() const [with T = c10::BFloat16]’:
/data/users/anijain/pytorch/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:376:3:   required from here
/data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:503:16: error: invalid conversion from ‘__m256 (*)(__m256)’ to ‘const __m256 (*)(__m256)’ [-fpermissive]
  503 |     return map(Sleef_coshf8_u10);
      |                ^~~~~~~~~~~~~~~~
      |                |
      |                __m256 (*)(__m256)
```

I have already sync'd the submodules. Any idea?

xuhancn commented:
> Not sure if related. I am on `main` branch (not on this PR) and seeing bulid failure like these
> 
> ```
> /data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h: In instantiation of ‘at::vec::AVX2::Vectorized<T> at::vec::AVX2::Vectorized16<T>::cosh() const [with T = c10::BFloat16]’:
> /data/users/anijain/pytorch/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:376:3:   required from here
> /data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:503:16: error: invalid conversion from ‘__m256 (*)(__m256)’ to ‘const __m256 (*)(__m256)’ [-fpermissive]
>   503 |     return map(Sleef_coshf8_u10);
>       |                ^~~~~~~~~~~~~~~~
>       |                |
>       |                __m256 (*)(__m256)
> ```
> 
> I have already sync'd the submodules. Any idea?
> 
> update - I fixed it in this PR - #122723. Not sure though.

Please wait for a while, My(this PR) will fix the sleef build issue. And current blocked on fb_code.

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
merge PR avoid code getting stale, @huydhn please update fb_code sleef to https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b, it should be works well. CC: @malfet

atalman commented:
@pytorchbot revert -m "Breaks internal build" -c ghfirst

atalman commented:
Please note multiple jobs are broken. here are some of the errors:
```
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:495:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_logf8_u10);
               ^~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:157:52: note: in instantiation of member function 'at::vec::Vectorized16<c10::BFloat16>::log' requested here
                return (x_vec / (kOneVec - x_vec)).log();
                                                   ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:495:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_logf8_u10);
               ^~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:157:52: note: in instantiation of member function 'at::vec::Vectorized16<c10::Half>::log' requested here
                return (x_vec / (kOneVec - x_vec)).log();
                                                   ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:510:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_sinhf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:371:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::BFloat16>::sinh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.sinh();});
                                                           ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:510:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_sinhf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:371:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::Half>::sinh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.sinh();});
                                                           ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:516:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_coshf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:380:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::BFloat16>::cosh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.cosh();});
                                                           ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:516:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_coshf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:380:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::Half>::cosh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.cosh();});
```

xuhancn committed:
try to fix fb_code sleef related build fail.

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot merge -i

huydhn commented:
@atalman I'm importing the sleef module to unblocking this internally, I have the diff to import sleef at D55567582.  We can rebase this on top to see if it fixes the build issue.
